File: C/exercise/RecentCounter.c

#include "RecentCounter.h"
#include <stdlib.h>

RecentCounter *recentCounterCreate()
{
	RecentCounter *obj = (RecentCounter *)malloc(sizeof(RecentCounter));
	obj->head = 0;
	obj->tail = 0;
	return obj;
}

int recentCounterPing(RecentCounter *obj, int t)
{
	obj->times[obj->tail] = t;
	obj->tail++;
	if (obj->tail >= MAX_SIZE) {
		obj->tail = 0;
	}
	while (obj->times[obj->head] < (t - 3000)) {
		obj->head++;
		if (obj->head >= MAX_SIZE) {
			obj->head = 0;
		}
	}
	return obj->tail - obj->head;
}

void recentCounterFree(RecentCounter *obj)
{
	free(obj);
}

/**
 * Your RecentCounter struct will be instantiated and called as such:
 * RecentCounter* obj = recentCounterCreate();
 * int param_1 = recentCounterPing(obj, t);

 * recentCounterFree(obj);
*/

File: Java/exercise/RecentCounter/RecentCounter.java

package exercise.RecentCounter;

public class RecentCounter {

	private int[] times = new int[10002];
	private int head = 0;
	private int tail = 0;

	public RecentCounter() {

	}

	public int ping(int t) {
		times[tail] = t;
		tail++;
		while (times[head] < t - 3000) {
			head++;
		}
		return tail - head;
	}
}

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter obj = new RecentCounter();
 * int param_1 = obj.ping(t);
 */
